src/util/ConfigManager/ConfigOptionProxy.h

```diff
@@ -0,0 +1,96 @@
+#include <type_traits>
+
+#include "util/ConfigManager/ConfigOption.h"
+#include "util/Exception.h"
+
+namespace ad_utility {
+namespace detail {
+/*
+@brief Implementation of proxy/reference to a `ConfigOption`. Also saves the
+type of value, that the configuration option holds. Needed, because without a
+`ConfigOptionProxy` and `ConstConfigOptionProxy`, people could easily discard
+the `const` from a `const ConfigOptionProxy` by copying it, to have non const
+access to the referenced config option.
+
+@tparam T The type of value, the later given config option will hold.
+@tparam ConfigOptionType The kind of config option, this proxy will reference
+to. Must be `ConfigOption`, or `const ConfigOption`.
+*/
+template <typename T, typename ConfigOptionType>
+requires ad_utility::isTypeContainedIn<T, ConfigOption::AvailableTypes> &&
+         (std::same_as<std::decay_t<ConfigOptionType>, ConfigOption> ||
+          std::same_as<std::decay_t<ConfigOptionType>, const ConfigOption>)
+class ConfigOptionProxyImplementation {
+  ConfigOptionType* option_ = nullptr;
+
+ public:
+  using Type = T;
+
+  // Custom constructor.
+  explicit ConfigOptionProxyImplementation(ConfigOptionType& opt)
+      : option_(&opt) {
+    // Make sure, that the given `ConfigOption` holds values of the right type.
+    if (!opt.template holdsType<T>()) {
+      throw std::runtime_error(absl::StrCat(
+          "Error while creating 'ConfigOptionProxy' for 'ConfigOption' '",
+          opt.getIdentifier(),
+          "': Mismatch between the template type of proxy and the type of "
+          "values, the configuration option can hold."));
+    }
+  }
+
+  // Get access to the configuration option, this is a proxy for. Const access
+  // only for the const version, to make the usage clearer.
+  ConfigOptionType& getConfigOption() const
+      requires std::is_const_v<ConfigOptionType> {
+    AD_CORRECTNESS_CHECK(option_ != nullptr);
+    return *option_;
+  }
+
+  ConfigOptionType& getConfigOption()
+      requires(!std::is_const_v<ConfigOptionType>) {
+    AD_CORRECTNESS_CHECK(option_ != nullptr);
+    return *option_;
+  }
+};
+
+}  // namespace detail
+
+// A const proxy/reference to a `ConfigOption`. Also saves the type of
+// value, that the configuration option holds.
+template <typename T>
+class ConstConfigOptionProxy
+    : public detail::ConfigOptionProxyImplementation<T, const ConfigOption> {
+  using Base = detail::ConfigOptionProxyImplementation<T, const ConfigOption>;
+
+ public:
+  explicit ConstConfigOptionProxy(const ConfigOption& opt) : Base(opt) {}
+};
+
+// A non const proxy/reference to a `ConfigOption`. Also saves the type of
+// value, that the configuration option holds.
+template <typename T>
+class NonConstConfigOptionProxy
+    : public detail::ConfigOptionProxyImplementation<T, ConfigOption> {
+  using Base = detail::ConfigOptionProxyImplementation<T, ConfigOption>;
+
+ public:
+  explicit NonConstConfigOptionProxy(ConfigOption& opt) : Base(opt) {}
```

Comment:
You can just call the mutable one `ConfigOptionProxy`

Reply:
Do you think, the name will be clear enough? I wanted to make it clear, that that proxy if only for non const.

Reply:
Yes, it is clear enough.

Reply:
Done, see [fea7699](https://github.com/ad-freiburg/qlever/pull/1055/commits/fea76991b9379471eb4daafc8eb61dfbf07c8354).